### PR TITLE
adapter: add plan insights for console hints

### LINKED ISSUE #26436: Add notice that, for each query, returns a new explain plan format for the console
Objective: Enable the console to warn users when the are running inefficient ad-hoc queries  https://github.com/MaterializeInc/console/issues/1977

Scope:
* Lay the foundation for this new "interface" for the console to get explain plan metadata for each query issued in the console shell. Design the format of the notice contents.
* For the first iteration, we only need to reveal which objects were read from persist (currently represented as ReadStorage in our textual explain plan representation, and `"access_strategy": "Persist"` in our JSON explain plan representation)

Proposed design: 
* Add a new session var `emit_explain_notice`.
    * We can bikeshed name, perhaps `emit_explain_summary_notice` to distinguish this unique format from the other ones?)
* When the session var is enabled, emit a notice for each statement in the query.
    * Open to discussion: should it instead be a single notice for the query with a list of explain summaries for each statement? One notice per statement allows us to eagerly emit the notice when beginning execution on the statement.
* Notice is in JSON format
* Todo: define the shape of this JSON object with the console team.
    * Probably something like `{"persist_accessed": [{"my_source"}, {"my_mat_view"}] }` will work to start
    * Note: we will _not_ make this part of our stable interface, so we are able to iterate on this with the console team, and do not need to be committed to the format

Open questions:
* What to do when `EXPLAIN PLAN` is not supported for the statement type, such as `SUBSCRIBE` or `INSERT`?

Future extension ideas:
* Whether a filter pushdown was applied for the persist access

### PR DESCRIPTION
**Pull request by benesch**
🚨 Heads up that I'm not going to be able to push this PR any further. 🚨 I'm hopeful it's nearly mergeable as is, but I'm chatting with @chaas about handing off the final bit of review/iteration to someone on the adapter team. —@benesch 

----

The two major commit titles/messages are reproduced below. There's a few more refactoring commits in the stack; see the individual commits and messages for details!

**adapter,sql: add EXPLAIN PLAN INSIGHTS statement**

`EXPLAIN PLAN INSIGHTS` is a new feature that produces key insights about
the plan for a `SELECT` statement.

In its initial form, its only consumer will be the console, which will
use it to power hints about e.g. missing indexes
(https://github.com/MaterializeInc/console/issues/1977). This allows us to keep the initial
implementation simple and focused--e.g., only the JSON format is
supported for now.

In the future, if the statement proves useful, we can consider
documenting the statement for end user use, and introducing a
human-readable textual format.

There's no feature flag for the statement, as the EXPLAIN statement is
not subject to backwards compatibility guarantees.

**adapter: add notice to emit plan insights**

Add a new `emit_plan_insights_notice` configuration parameter which, for
every `SELECT` statement, emits the output of `EXPLAIN PLAN INSIGHTS` as
a notice alongside the results.

This will allow the console's SQL shell to inspect the insights and
render hints about the executing SQL statement as appropriate
(https://github.com/MaterializeInc/console/issues/1977).

Fix #26436.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Go commit by commit. Don't be scared by the diffstat, either: it's mostly explain plans in tests.

@mjibson @aalexandrov I've done my best to hew to the `EXPLAIN` code's norms, but the code has changed quite a bit since I last looked at it and I'm sure I didn't get everything quite right.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by SangJunBak** [comment]
I noticed the payload doesn't contain what kind of notice it is via a code. Is this intentional?

**Comment by benesch** [comment]
The code should be there! It's presently `MZ001`. It just doesn't show in `psql` by default. Here's an updated example which displays the code.

```
materialize=> set emit_plan_insights_notice = true;
SET
materialize=> \set VERBOSITY 'verbose'
materialize=> select * from t; select * from t;
NOTICE:  MZ001: {
  "plans": {
    "raw": "Finish output=[#0]\n  Get materialize.public.t\n",
    "optimized": {
      "global": "t46:\n  Finish output=[#0]\n    ArrangeBy keys=[[#0]]\n      ReadGlobalFromSameDataflow t45\n\nt45:\n  ReadStorage materialize.public.t\n"
    }
  },
  "insights": {
    "imports": {
      "u20": {
        "name": {
          "database": "materialize",
          "schema": "public",
          "item": "t"
        },
        "type": "storage"
      }
    }
  }
}
 a 
---
(0 rows)

NOTICE:  MZ001: {
  "plans": {
    "raw": "Finish output=[#0]\n  Get materialize.public.t\n",
    "optimized": {
      "global": "t49:\n  Finish output=[#0]\n    ArrangeBy keys=[[#0]]\n      ReadGlobalFromSameDataflow t48\n\nt48:\n  ReadStorage materialize.public.t\n"
    }
  },
  "insights": {
    "imports": {
      "u20": {
        "name": {
          "database": "materialize",
          "schema": "public",
          "item": "t"
        },
        "type": "storage"
      }
    }
  }
}
 a 
---
(0 rows)

```

**Comment by SangJunBak** [comment]
> The code should be there! It's presently `MZ001`. It just doesn't show in `psql` by default. Here's an updated example which displays the code.
> 
> ```
> materialize=> set emit_plan_insights_notice = true;
> SET
> materialize=> \set VERBOSITY 'verbose'
> materialize=> select * from t; select * from t;
> NOTICE:  MZ001: {
>   "plans": {
>     "raw": "Finish output=[#0]\n  Get materialize.public.t\n",
>     "optimized": {
>       "global": "t46:\n  Finish output=[#0]\n    ArrangeBy keys=[[#0]]\n      ReadGlobalFromSameDataflow t45\n\nt45:\n  ReadStorage materialize.public.t\n"
>     }
>   },
>   "insights": {
>     "imports": {
>       "u20": {
>         "name": {
>           "database": "materialize",
>           "schema": "public",
>           "item": "t"
>         },
>         "type": "storage"
>       }
>     }
>   }
> }
>  a 
> ---
> (0 rows)
> 
> NOTICE:  MZ001: {
>   "plans": {
>     "raw": "Finish output=[#0]\n  Get materialize.public.t\n",
>     "optimized": {
>       "global": "t49:\n  Finish output=[#0]\n    ArrangeBy keys=[[#0]]\n      ReadGlobalFromSameDataflow t48\n\nt48:\n  ReadStorage materialize.public.t\n"
>     }
>   },
>   "insights": {
>     "imports": {
>       "u20": {
>         "name": {
>           "database": "materialize",
>           "schema": "public",
>           "item": "t"
>         },
>         "type": "storage"
>       }
>     }
>   }
> }
>  a 
> ---
> (0 rows)
> ```

@benesch ah I see! It seems `\set verbosity` is a psql specific command. Is there a setting I can change to expose this via the websocket protocol?

**Comment by SangJunBak** [comment]
I noticed something like `SELECT * FROM mz_sources` gives an order like:
```
Notice
CommandStarting
Rows
Row
...
CommandComplete
``` 
but `select 1/0` gives the notice in this order:
```
CommandStarting
Error
Notice
```

Will we ever get a case where we have a successful query after an error? i.e. `select 1/0; SELECT * FROM mz_sources;` would give
```
CommandStarting
Error
Notice

Notice
CommandStarting
Rows
Row
...
CommandComplete
ReadyForQuery
```

If we do, it'll be hard. to associate what notices belong to which command.

**Comment by benesch** [comment]
> @benesch ah I see! It seems `\set verbosity` is a psql specific command. Is there a setting I can change to expose this via the websocket protocol?

Ah, sorry! I guess notice codes just weren't included in the WebSocket protocol. I've pushed up a fix for that in the latest revision of this PR.

> Realized the placement of the notices are ambiguous.

Whoops, yes, thanks! Fixed in the latest revision. The plan insights notice should now appear, when supported, immediately before the `CommandStarting` message for the corresponding command. (I'd prefer for it to be _between_ the `CommandStarting` and `CommandComplete`, but for annoying code structure reasons I can't easily fix this.)

File: src/adapter/src/coord/sequencer/inner/peek.rs [hunk 1]
```diff
@@ -75,14 +76,25 @@ impl Coordinator {
     ) {
         event!(Level::TRACE, plan = format!("{:?}", plan));
 
+        let explain_ctx = match ctx.session().vars().emit_plan_insights_notice() {
+            true => {
+                let broken = false;
```
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner/peek.rs:81]: Note that this means that any regular tracing spans / events emitted while the optimizer is running will be suppressed while `emit_plan_insights_notice` is `true`.

This is because you will be hitting the `else` branch in `OptimizerTrace::new(...)` constructor, which creates the subscriber chain rooted at `tracing_subscriber::registry()`. If you want to maintain the tracing capabilities if notices are turned on I suggest to set `broken = true`. We had some issues in the past where `broken = true` and using the `let subscriber = DelegateSubscriber::default()` was causing memory leaks, but we have not been seeing this recently.

File: src/sql-parser/src/ast/defs/statement.rs [hunk 2]
```diff
@@ -3752,6 +3754,7 @@ impl ExplainStage {
             Self::GlobalPlan => Some(Global.path()),
             Self::PhysicalPlan => Some(Physical.path()),
             Self::Trace => None,
+            Self::PlanInsights => None,
```
Comment (aalexandrov) [inline src/sql-parser/src/ast/defs/statement.rs:3757]: This is probably OK for an MVP, but might be impacting performance in regular `peek` mode when `emit_plan_insights_notice = true` because it will cause the `trace_plan(...)` calls in our optimizer to always create an entry. In other words, we will collect / clone every intermediate result in the optimizer pipeline in the accumulated `OptimizerTrace` instance, even though at the end we will extract only the `RawPlan` and `GlobalPlan` stages.

File: src/adapter/src/explain/optimizer_trace.rs [hunk 3]
```diff
@@ -133,24 +138,62 @@ impl OptimizerTrace {
...
+                        "text": text_plan,
+                        "json": json_plan,
+                    }))
+                };
+
+                let output = serde_json::json!({
+                    "plans": {
+                        "raw": get_plan(NamedPlan::Raw)?,
+                        "optimized": {
+                            "global": get_plan(NamedPlan::Global)?,
```
Comment (aalexandrov) [inline src/adapter/src/explain/optimizer_trace.rs:172]: Are we not interested in the `optimize/fast_path` result here as well? A lot of `SELECT` queries are executed through that path and per default `EXPLAIN <SELECT>` would render the plan located at `optimize/fast_path` instead of `optimize/global`.

**Review by aalexandrov — APPROVED** [review]
LGTM, except some minor nits.

I've left some comments in places where I think we might hit performance issues or lose other tracing capabilities, but they are certainly not blocking for an MVP.

Also, note that if the primary goal is to collect and emit certain notices as a by-product of running an optimization pass as a psql notice, we can also do that outside of the `EXPLAIN` framework. 

Instead, we can synthesize and collect the required "insights" directly in the optimizer pipelines as an extra field in the `DataflowMetainfo` instance, which is now emitted as a by-product of each optimizer pass (for example search for `DataflowMetainfo` in `src/adapter/src/optimize/peek.rs`). I don't know what exactly the cited complications were to add those notices for all statement types with the `EXPLAIN`-based method, but maybe this will make the implementation easier.

File: src/adapter/src/coord/sequencer/inner/peek.rs (hunk 1 above)
Comment (benesch) [reply src/adapter/src/coord/sequencer/inner/peek.rs:81]: Isn't `broken = false` what `EXPLAIN ... SELECT` uses by default? Was just trying to match that standard flow!

File: src/adapter/src/explain/optimizer_trace.rs (hunk 3 above)
Comment (benesch) [reply src/adapter/src/explain/optimizer_trace.rs:172]: Ah, that's a good idea. @mjibson do you mind stitching this in as `plans.optimized.fast_path`?

File: src/sql-parser/src/ast/defs/statement.rs (hunk 2 above)
Comment (benesch) [reply src/sql-parser/src/ast/defs/statement.rs:3757]: Makes sense. How hard would it be to adjust the optimizer to allow ORing together multiple path filters?

**Comment by benesch** [comment]
Synced a bit with @SangJunBak and @chaas today in the office today. Apologies that I don't have the space in my schedule this week for all of us to sync live on this. A few thoughts on the approach in this PR vs the optimizer notices framework below.

> Also, note that if the primary goal is to collect and emit certain notices as a by-product of running an optimization pass as a psql notice, we can also do that outside of the `EXPLAIN` framework.
> 
> Instead, we can synthesize and collect the required "insights" directly in the optimizer pipelines as an extra field in the `DataflowMetainfo` instance, which is now emitted as a by-product of each optimizer pass (for example search for `DataflowMetainfo` in `src/adapter/src/optimize/peek.rs`). I don't know what exactly the cited complications were to add those notices for all statement types with the `EXPLAIN`-based method, but maybe this will make the implementation easier.

> Talked with Alexander about this offline. Given it's not a big lift on the Console side, it makes sense to merge Nikhil's implementation but eventually port this to the optimizer framework and have it generate the message rather than the Console.

It might make sense in the (very) far future to let the optimizer handle the construction of the notice messages, but in the short/medium term I think we very much want to leave the construction of the notice itself to the console. We'll want to do a a lot of tinkering and iterating to get the wording, layout, CTA, etc. of these notices just right, and that's much easier to do if the notice generation lives entirely with the console. We can ship changes to the console multiple times a day, while shipping changes to the optimizer requires waiting 1-2 weeks for a release to go out.

I think of the options here as lying on a spectrum. The left end of the spectrum is "let the optimizer generate everything about the notices" and the right end of the spectrum is "have the console parse the EXPLAIN plan JSON and derive everything itself." The left end is painful because iteration takes weeks, and the right end is painful because the EXPLAIN JSON is hard to parse and isn't stable.

The approach in this PR is intentionally in the middle of the spectrum. The database helps the console out by deriving some insights about the explain plan and presenting them in a stable* format, which spares the console from having to parse the complicated and unstable JSON format. But the console is also free to use the insights to present a variety of different notices as it chooses; it's not locked in to the specific wording that the optimizer is using in that version of Materialize.

<small>*We're not committed to backcompat for the plan insights JSON, but I say "stable" because we'd make sure not to break the console when evolving the plan insights JSON's format.</small>

If the notices stabilize over time, we may start to think about moving to a unified world where the optimizer is in charge of all notice generation. 

> This is because as we increase the complexities of these notices, such as if a user put a wrong temporal filter for an MFP pushdown, it's must easier for mz to figure this out rather than the Console.

This is actually exactly what I'm envisioning insights being useful for! We can easily add additional insights about why MFP pushdown did or did not apply, without being prescriptive about how the console chooses to render those insights into notices.

File: src/adapter/src/explain/optimizer_trace.rs (hunk 3 above)
Comment (mjibson) [reply src/adapter/src/explain/optimizer_trace.rs:174]: Done.

**Comment by mjibson** [comment]
Rebased, fixed tests, made the addition for fast_path in the json output. Leaving the other things that were described as being ok for MVP and can come later.

File: src/adapter/src/coord/sequencer/inner/peek.rs (hunk 1 above)
Comment (aalexandrov) [reply src/adapter/src/coord/sequencer/inner/peek.rs:81]: It is, but it currently has the somewhat unpleasant behavior of hiding all `tracing::span!` and `tracing::event!` calls that happen when you call
```rust
let _dispatch_guard = explain_ctx.dispatch_guard();
```
in the spawned optimization tasks because the `&dispatch` that is temporary set with
```rust
tracing::dispatcher::set_default(&dispatch)
```
is not rooted at the `DelegateSubscriber::default()` when `broken` is `false`. We did this because at some point I think there was a bug in `tokio::tracing` that was causing memory leaks (which so far were only visible when you run `EXPLAIN` in a hot loop like in the SQLSmith tests).

So far getting limited tracing support here had limited impact because it only affected `EXPLAIN` paths. If you really wanted to see `tracing::event!` lines for an `EXPLAIN <explainee>` statement (as opposed to just running `<explainee>` directly), you could do
```sql
ALTER SYSTEM SET enable_explain_broken = true;
```
and then type
```sql
EXPLAIN BROKEN <explainee> ...
```

However, because `ExplainContext::PlanInsightsNotice` is now set whenever `emit_plan_insights_notice` is set, it means that this behavior will extend to all `SELECT` queries that are executed from the console.

So my suggestion is to:

1. Set `let broken = true;` here for now in order to retain visibility in the console.
2. I'll prepare a PR (or a commit in this PR) that introduces a feature flag that forces `OptimizerTrace::new(...)` to be called with `true` when set, set the feature flag default to `true` everywhere, and do some local tests. If there are no issues with the next release we can remove the flag and the conditional in `OptimizerTrace::new(...)` with the next release.

File: src/sql-parser/src/ast/defs/statement.rs (hunk 2 above)
Comment (aalexandrov) [reply src/sql-parser/src/ast/defs/statement.rs]: I added a commit to this PR that fixes the issue.

**Comment by aalexandrov** [comment]
Rebased, fixed broken tests, and added two commits that should close the discussion [here](https://github.com/MaterializeInc/materialize/pull/26504#discussion_r1558780176) and [here](https://github.com/MaterializeInc/materialize/pull/26504#discussion_r1557833471). I thing if all tests pass this should be ready to be merged.

**Comment by aalexandrov** [comment]
I had to change the default value of `enable_broken_optimizer_trace` to `false` because the memory leaks [described here](https://github.com/MaterializeInc/materialize/pull/26504#discussion_r1559071250) are still present (see [this Buildkite failure](https://buildkite.com/materialize/nightly/builds/7339#018ec82b-1f32-4bd8-acd6-0893c909c662)).

The flag will give you the option to decide whether tokio traces and logs that happen within the scope of a `emit_plan_insights_notice = true` session should be visible at the expense of a very slow memory leak until we fix it.

**Comment by aalexandrov** [comment]
OK, this is _super weird_... 

This loop is not causing a leak, so it is probably even safe to hard-code `let broken = true` in the new case added by this PR...

```bash
while true; do 
  psql --set verbosity=verbose \
    -c 'set emit_plan_insights_notice = true' \
    -c 'select count(*) from bids b left join users u on (b.buyer = u.id)' > /dev/null
done
```

However, this loop

```bash
while true; do
  psql -c 'explain plan insights as json for select count(*) from bids b left join users u on (b.buyer = u.id)' > /dev/null
done
```

is not causing a leak when `enable_broken_optimzier_trace = false` (which is consistent with the current behavior in `main`), but is causing a leak when `enable_broken_optimzier_trace = true`. I'll open a separate issue to investigate.

**Comment by mjibson** [comment]
@aalexandrov I'm lightly leaning toward merging this pr as is with `broken = false` until 26569 is more understood. Thoughts?

**Comment by benesch** [comment]
> @aalexandrov I'm lightly leaning toward merging this pr as is with `broken = false` until 26569 is more understood. Thoughts?

That's the current state of the PR, right? My read was that @aalexandrov was on board with this, since he filed #26569 to follow up.

**Comment by mjibson** [comment]
Right. Ok, merging then.

**Comment by benesch** [comment]
Hooray! Thanks very much for the assist on this, @mjibson and @aalexandrov!

**Comment by danhhz** [comment]
Any chance this broke main? The last couple builds have the same failure and this is the first PR it happened to https://buildkite.com/materialize/test/builds/80410

**Comment by benesch** [comment]
Looks like yes. Quick PR: https://github.com/MaterializeInc/materialize/pull/26577
